io/iommu/iommu_tests.py

```diff
@@ -372,8 +377,9 @@ def test_reset_rescan(self):
             self.rescan(pci_addr)
             # check the device for default state
             self.check(def_dom, pci_addr, driver)
+        self.check_dmesg()
 
-    def tearDown(self):
+    def check_dmesg(self):
```

Comment:
@dhsrivas why do you want to remove the tearDown()?
you can add the code to check_dmesg() and handle same from tearDown.
I suggest to keep the framework flow which helps for proper entry and exit of the code

Reply:
#avocado run io/iommu/iommu_tests.py -p pci_devices="0000:03:00.3" --test-runner runner
JOB ID     : a7bb3169d2735f5e69e201ae82188a351a1b44b3
JOB LOG    : /root/avocado/job-results/job-2023-06-06T08.34-a7bb316/job.log
 (1/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_bind: PASS (5.27 s)
 (2/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_bind: PASS (62.19 s)
 (3/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_ntimes_bind: PASS (62.19 s)
 (4/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_bind_rescan: PASS (5.40 s)
 **(5/6) io/iommu/iommu_tests.py:IommuTest.test_unbind_changedomain_bind_rescan: ERROR: Kernel Errors: [Tue Jun  6 08:36:43 2023] hub 1-0:1.0: activate --> -19\n (62.46 s)**
 (6/6) io/iommu/iommu_tests.py:IommuTest.test_reset_rescan: PASS (0.44 s)
RESULTS    : PASS 5 | ERROR 1 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 198.46 s

@Naresh-ibm I tried to call check_dmesg() from tearDown() instead of putting check_dmesg() after every test. (code snippet added below). If you observe the output still its not resolving the issue, i.e self.fail is still coming in ERROR instead of FAIL. 

```
def check_dmesg(self):
        """
        Checks for any error or failure messages in dmesg after test
        """
        cmd = "dmesg -T --level=alert,crit,err,warn"
        out = process.run(cmd, ignore_status=True, shell=True, sudo=True)
        output = out.stdout_text
        if output:
            self.fail(f"Kernel Errors: {output}")

def tearDown(self):
        self.check_dmesg()
```

Reply:
then something is not right in self.fail?
please re-check in other test cases, cross verify your usage, and you can report an issue in avocado for same to get rectified if nothing is is wrong in your code.

I feel the the test is erroring out before the output check itself, that could be the reason for erroring out instead of failure.
check if any command is returns a non-zero value.

Reply:
> I feel the the test is erroring out before the output check itself, that could be the reason for erroring out instead of failure.

No, in that case it would have error out even with (instead of tearDown) using check_dmesg() after each test)
```
def check_dmesg(self):
        """
        Checks for any error or failure messages in dmesg after test
        """
        cmd = "dmesg -T --level=alert,crit,err,warn"
        out = process.run(cmd, ignore_status=True, shell=True, sudo=True)
        output = out.stdout_text
        if output:
            self.fail(f"Kernel Errors: {output}")
```
Also adding snippet of debug.log to show in case of error out it is coming to self.fail part of teraDown
```
08:37:31 ERROR|
08:37:31 ERROR| Reproduced traceback from: /usr/local/lib/python3.10/dist-packages/avocado_framework-92.0-py3.10.egg/avocado/core/test.py:819
08:37:31 ERROR| Traceback (most recent call last):
08:37:31 ERROR|   File "/root/avocado-misc-tests/io/iommu/iommu_tests.py", line 393, in tearDown
08:37:31 ERROR|     self.check_dmesg()
08:37:31 ERROR|   File "/root/avocado-misc-tests/io/iommu/iommu_tests.py", line 390, in check_dmesg
08:37:31 ERROR|     **self.fail(f"Kernel Errors: {output}")**
08:37:31 ERROR|   File "/usr/local/lib/python3.10/dist-packages/avocado_framework-92.0-py3.10.egg/avocado/core/test.py", line 979, in fail
08:37:31 ERROR|     raise exceptions.TestFail(message)
08:37:31 ERROR| avocado.core.exceptions.TestFail: Kernel Errors: [Tue Jun  6 08:36:43 2023] hub 1-0:1.0: activate --> -19

```

> please re-check in other test cases, cross verify your usage

Other tests seems to be using the self.fail the same way.
Also i dont find any test using self.fail in setUp or tearDown section.

Reply:
I feel PR is good as TearDown only can be used to revert any setup done in the SetUp section and test  decisions like Fail/PaaS/Error decisions only be taken care of in the test